plugins/hls-refine-imports-plugin/src/Ide/Plugin/RefineImports.hs

```diff
@@ -0,0 +1,271 @@
+  | TextDocumentIdentifier {_uri} <- docId,
+    Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri = liftIO $
+    do
+      pm <- runIde ideState $ use GetParsedModule nfp
+      let insideImport = case pm of
+            Just ParsedModule {pm_parsed_source}
+              | locImports <- hsmodImports (unLoc pm_parsed_source),
+                rangesImports <- map getLoc locImports ->
+                any (within range) rangesImports
+            _ -> False
+      if not insideImport
+        then return (Right (List []))
+        else do
+          mbRefinedImports <- runIde ideState $ use RefineImports nfp
+          let edits =
+                [ e
+                | Just (RefineImportsResult result) <- [mbRefinedImports]
+                , (imp, Just refinedImports) <- result
+                , Just e <- [mkExplicitEdit zeroMapping imp refinedImports]
+                ]
+              caExplicitImports = InR CodeAction {..}
+              _title = "Refine all imports"
+              _kind = Just CodeActionQuickFix
+              _command = Nothing
+              _edit = Just WorkspaceEdit
+                {_changes, _documentChanges, _changeAnnotations}
+              _changes = Just $ HashMap.singleton _uri $ List edits
+              _documentChanges = Nothing
+              _diagnostics = Nothing
+              _isPreferred = Nothing
+              _disabled = Nothing
+              _xdata = Nothing
+              _changeAnnotations = Nothing
+          return $ Right $ List [caExplicitImports | not (null edits)]
+  | otherwise =
+    return $ Right $ List []
+
+--------------------------------------------------------------------------------
+
+data RefineImports = RefineImports
+  deriving (Show, Generic, Eq, Ord)
+
+instance Hashable RefineImports
+instance NFData RefineImports
+instance Binary RefineImports
+type instance RuleResult RefineImports = RefineImportsResult
+
+newtype RefineImportsResult = RefineImportsResult
+  {getMinimalImportsResult :: [(LImportDecl GhcRn, Maybe T.Text)]}
+
+instance Show RefineImportsResult where show _ = "<refineImportsResult>"
+instance NFData RefineImportsResult where rnf = rwhnf
+
+refineImportsRule :: Rules ()
+refineImportsRule = define $ \RefineImports nfp -> do
+  -- Get the typechecking artifacts from the module
+  tmr <- use TypeCheck nfp
+  -- We also need a GHC session with all the dependencies
+  hsc <- use GhcSessionDeps nfp
+
+  -- 2 layer map ModuleName -> ModuleName -> [Avails] (exports)
+  import2Map <- do
+    -- first layer is from current(editing) module to its imports
+    ImportMap currIm <- use_ GetImportMap nfp
+    forM currIm $ \path -> do
+      -- second layer is from the imports of first layer to their imports
+      ImportMap importIm <- use_ GetImportMap path
+      forM importIm $ \imp_path -> do
+        imp_tmr <- use_ TypeCheck imp_path
+        return $ tcg_exports $ tmrTypechecked imp_tmr
```

Comment:
This construction is inaccurate. Counterexample:

```
module A where
import B (b)
a = b
```
```
module B(b) where
import qualified C
b = not C.b
```
```
module C where
b = False
```

This plugin will refine A to:

```
module A where
import C (b)
a = b
```

Reply:
Nice small example to be added as a test :wink:

Reply:
will do thanks!